File: src/ir.zig
//! Ansilust IR - Core intermediate representation
//!
//! Implements the Cell Grid IR (Approach 1) from IR-RESEARCH.md
//! Designed to be compatible with OpenTUI's OptimizedBuffer

const std = @import("std");

/// Main IR structure
pub const AnsilustIR = struct {
    allocator: std.mem.Allocator,

    // Canvas dimensions
    width: u32,
    height: u32,

    // Cell grid (flattened: width × height)
    cells: []Cell,

    // Reference-counted style table (Ghostty pattern)
    style_table: std.ArrayList(Style),

    // Multi-codepoint grapheme storage
    grapheme_map: std.AutoHashMap(u32, []u32),

    // Resources
    palette: PaletteType,
    font: FontInfo,

    // Metadata
    sauce: ?SauceRecord,
    source_format: SourceFormat,

    // Rendering hints
    ice_colors: bool,
    letter_spacing: u8, // 8 or 9
    aspect_ratio: ?f32, // 1.35 for DOS

    pub fn init(allocator: std.mem.Allocator, width: u32, height: u32) !AnsilustIR {
        const size = width * height;
        const cells = try allocator.alloc(Cell, size);

        // Initialize with default cells
        const default_cell = Cell{
            .char = ' ',
            .style_id = 0,
            .flags = CellFlags{},
        };

        for (cells) |*cell| {
            cell.* = default_cell;
        }

        var style_table: std.ArrayList(Style) = .empty;

        const default_style = Style{
            .fg = Color{ .palette = 7 }, // White
            .bg = Color{ .palette = 0 }, // Black
            .underline_color = null,
            .attributes = Attributes{},
            .hyperlink = null,
        };
        try style_table.append(allocator, default_style);

        return AnsilustIR{
            .allocator = allocator,
            .width = width,
            .height = height,
            .cells = cells,
            .style_table = style_table,
            .grapheme_map = std.AutoHashMap(u32, []u32).init(allocator),
            .palette = .ansi,
            .font = FontInfo{
                .id = "cp437",
                .width = 8,
                .height = 16,
                .embedded = null,
            },
            .sauce = null,
            .source_format = .ansi,
            .ice_colors = false,
            .letter_spacing = 8,
            .aspect_ratio = null,
        };
    }

    pub fn deinit(self: *AnsilustIR) void {
        self.allocator.free(self.cells);
        self.style_table.deinit(self.allocator);

        // Free grapheme map entries
        var it = self.grapheme_map.iterator();
        while (it.next()) |entry| {
            self.allocator.free(entry.value_ptr.*);
        }
        self.grapheme_map.deinit();

        if (self.sauce) |sauce| {
            sauce.deinit(self.allocator);
        }
    }

    pub fn getCell(self: *const AnsilustIR, x: u32, y: u32) ?*const Cell {
        if (x >= self.width or y >= self.height) return null;
        const index = y * self.width + x;
        return &self.cells[index];
    }

    pub fn setCell(self: *AnsilustIR, x: u32, y: u32, cell: Cell) bool {
        if (x >= self.width or y >= self.height) return false;
        const index = y * self.width + x;
        self.cells[index] = cell;
        return true;
    }

    pub fn getStyle(self: *const AnsilustIR, style_id: u16) ?*const Style {
        if (style_id >= self.style_table.items.len) return null;
        return &self.style_table.items[style_id];
    }

    pub fn addStyle(self: *AnsilustIR, style: Style) !u16 {
        // Check if style already exists (reference counting)
        for (self.style_table.items, 0..) |existing, id| {
            if (std.meta.eql(existing, style)) {
                return @intCast(id);
            }
        }

        // Add new style
        try self.style_table.append(self.allocator, style);
        return @intCast(self.style_table.items.len - 1);
    }
};

/// Individual cell in the grid
pub const Cell = packed struct {
    char: u32, // Unicode codepoint or CP437 code
    style_id: u16, // Index into style_table
    flags: CellFlags,
};

/// Cell flags (inspired by Ghostty)
pub const CellFlags = packed struct {
    wide_char: bool = false, // Double-width character
    spacer_tail: bool = false, // Second half of wide char
    spacer_head: bool = false, // Wrapped wide char marker
    soft_wrap: bool = false, // Line soft-wraps to next
    protected: bool = false, // Write-protected cell
    _padding: u3 = 0, // Padding to full byte
};

/// Style information (reference-counted)
pub const Style = struct {
    fg: Color,
    bg: Color,
    underline_color: ?Color,
    attributes: Attributes,
    hyperlink: ?u32,
};

/// Color representation
pub const Color = union(enum) {
    none, // Terminal default (not black!)
    palette: u8, // Index 0-255
    rgb: RGB, // 24-bit true color
};

/// RGB color
pub const RGB = struct {
    r: u8,
    g: u8,
    b: u8,
};

/// Text attributes (Ghostty-inspired with BBS extensions)
pub const Attributes = packed struct {
    bold: bool = false,
    faint: bool = false,
    italic: bool = false,
    underline: bool = false,
    underline_double: bool = false,
    underline_curly: bool = false,
    blink: bool = false,
    reverse: bool = false,
    invisible: bool = false,
    strikethrough: bool = false,
    overline: bool = false,
    _padding: u5 = 0, // Padding to 16 bits
};

/// Palette type
pub const PaletteType = union(enum) {
    ansi,
    vga,
    workbench,
    custom: Palette,
};

/// Custom palette definition
pub const Palette = struct {
    colors: []RGB,

    pub fn deinit(self: Palette, allocator: std.mem.Allocator) void {
        allocator.free(self.colors);
    }
};

/// Font information
pub const FontInfo = struct {
    id: []const u8,
    width: u8,
    height: u8,
    embedded: ?BitmapFont,
};

/// Bitmap font data
pub const BitmapFont = struct {
    width: u8,
    height: u8,
    char_count: u16,
    data: []u8,

    pub fn deinit(self: BitmapFont, allocator: std.mem.Allocator) void {
        allocator.free(self.data);
    }
};

/// SAUCE metadata record
pub const SauceRecord = struct {
    title: []const u8,
    author: []const u8,
    group: []const u8,
    date: []const u8,
    columns: u16,
    rows: u16,
    flags: SauceFlags,
    font_name: []const u8,
    comments: [][]const u8,

    pub fn deinit(self: SauceRecord, allocator: std.mem.Allocator) void {
        allocator.free(self.title);
        allocator.free(self.author);
        allocator.free(self.group);
        allocator.free(self.date);
        allocator.free(self.font_name);
        for (self.comments) |comment| {
            allocator.free(comment);
        }
        allocator.free(self.comments);
    }
};

/// SAUCE flags
pub const SauceFlags = packed struct {
    ice_colors: bool = false,
    _reserved1: bool = false,
    letter_spacing_9bit: bool = false,
    aspect_ratio: bool = false,
    _padding: u4 = 0,
};

/// Source format type
pub const SourceFormat = enum {
    ansi,
    binary,
    pcboard,
    xbin,
    tundra,
    artworx,
    icedraw,
    utf8ansi,
};

// Tests
test "create IR" {
    const allocator = std.testing.allocator;

    var ir = try AnsilustIR.init(allocator, 80, 25);
    defer ir.deinit();

    try std.testing.expectEqual(@as(u32, 80), ir.width);
    try std.testing.expectEqual(@as(u32, 25), ir.height);
    try std.testing.expectEqual(@as(usize, 80 * 25), ir.cells.len);
}

test "cell access" {
    const allocator = std.testing.allocator;

    var ir = try AnsilustIR.init(allocator, 80, 25);
    defer ir.deinit();

    // Set cell
    const cell = Cell{
        .char = 'A',
        .style_id = 0,
        .flags = CellFlags{},
    };
    try std.testing.expect(ir.setCell(0, 0, cell));

    // Get cell
    const retrieved = ir.getCell(0, 0).?;
    try std.testing.expectEqual(@as(u32, 'A'), retrieved.char);

    // Out of bounds
    try std.testing.expectEqual(@as(?*const Cell, null), ir.getCell(100, 100));
    try std.testing.expect(!ir.setCell(100, 100, cell));
}

test "style table" {
    const allocator = std.testing.allocator;

    var ir = try AnsilustIR.init(allocator, 80, 25);
    defer ir.deinit();

    const style = Style{
        .fg = Color{ .rgb = RGB{ .r = 255, .g = 0, .b = 0 } },
        .bg = Color.none,
        .underline_color = null,
        .attributes = Attributes{ .bold = true },
        .hyperlink = null,
    };

    const id1 = try ir.addStyle(style);
    const id2 = try ir.addStyle(style); // Should reuse

    try std.testing.expectEqual(id1, id2); // Reference counting
    try std.testing.expectEqual(@as(usize, 2), ir.style_table.items.len); // Default + new
}

test "cell flags" {
    const flags = CellFlags{
        .wide_char = true,
        .soft_wrap = true,
    };
    try std.testing.expect(flags.wide_char);
    try std.testing.expect(flags.soft_wrap);
    try std.testing.expect(!flags.spacer_tail);
}

File: src/ir.rs
//! Core IR data structures for ansilust
//! 
//! Implements the Cell Grid IR (Approach 1) from IR-RESEARCH.md

use bitflags::bitflags;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Main intermediate representation structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnsilustIR {
    /// IR format version
    pub version: String,
    
    /// Canvas dimensions
    pub width: u32,
    pub height: u32,
    
    /// Flattened cell grid (width × height)
    pub cells: Vec<Cell>,
    
    /// Reference-counted style table (Ghostty pattern)
    pub style_table: Vec<Style>,
    
    /// Multi-codepoint grapheme storage
    pub grapheme_map: HashMap<u32, Vec<u32>>,
    
    /// Resources
    pub palette: PaletteType,
    pub font: FontInfo,
    
    /// Metadata
    pub sauce: Option<SauceRecord>,
    pub source_format: SourceFormat,
    
    /// Rendering hints (from SAUCE or command-line)
    pub ice_colors: bool,
    pub letter_spacing: u8,  // 8 or 9
    pub aspect_ratio: Option<f32>,  // 1.35 for DOS
    
    /// Animation frames (optional)
    pub frames: Option<Vec<Frame>>,
}

/// Individual cell in the grid
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Cell {
    /// Unicode codepoint or CP437 character code
    pub char: u32,
    
    /// Index into style_table
    pub style_id: u16,
    
    /// Cell flags (wide char, wrap, etc.)
    pub flags: CellFlags,
}

bitflags! {
    /// Cell flags (inspired by Ghostty)
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
    pub struct CellFlags: u8 {
        /// Double-width character (CJK, emoji)
        const WIDE_CHAR      = 0b0000_0001;
        /// Second half of wide char
        const SPACER_TAIL    = 0b0000_0010;
        /// Wrapped wide char marker
        const SPACER_HEAD    = 0b0000_0100;
        /// Line soft-wraps to next row
        const SOFT_WRAP      = 0b0000_1000;
        /// Write-protected cell
        const PROTECTED      = 0b0001_0000;
    }
}

/// Style information (reference-counted)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Style {
    pub fg: Color,
    pub bg: Color,
    /// Separate underline color (can differ from fg!)
    pub underline_color: Option<Color>,
    pub attributes: Attributes,
    /// Optional hyperlink reference
    pub hyperlink: Option<u32>,
}

/// Color representation (supports classic and modern)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Color {
    /// Terminal default (not black!)
    None,
    /// Palette index (0-255)
    Palette(u8),
    /// 24-bit true color
    RGB(u8, u8, u8),
}

bitflags! {
    /// Text attributes (Ghostty-inspired with BBS extensions)
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
    pub struct Attributes: u16 {
        const BOLD              = 0b0000_0000_0000_0001;
        const FAINT             = 0b0000_0000_0000_0010;
        const ITALIC            = 0b0000_0000_0000_0100;
        const UNDERLINE         = 0b0000_0000_0000_1000;
        const UNDERLINE_DOUBLE  = 0b0000_0000_0001_0000;
        const UNDERLINE_CURLY   = 0b0000_0000_0010_0000;
        const BLINK             = 0b0000_0000_0100_0000;
        const REVERSE           = 0b0000_0000_1000_0000;
        const INVISIBLE         = 0b0000_0001_0000_0000;
        const STRIKETHROUGH     = 0b0000_0010_0000_0000;
        const OVERLINE          = 0b0000_0100_0000_0000;
    }
}

/// Palette type
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PaletteType {
    /// Standard ANSI palette
    ANSI,
    /// Standard VGA palette
    VGA,
    /// Amiga Workbench palette
    Workbench,
    /// Custom palette (16 or 256 colors)
    Custom(Palette),
}

/// Custom palette definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Palette {
    pub colors: Vec<RGB>,
}

/// RGB color
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct RGB {
    pub r: u8,
    pub g: u8,
    pub b: u8,
}

/// Font information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FontInfo {
    /// Font ID (e.g., "cp437", "topaz")
    pub id: Option<String>,
    /// Character width (8 or 9)
    pub width: u8,
    /// Character height (8, 16, 32)
    pub height: u8,
    /// Embedded bitmap font data (for XBin, ArtWorx)
    pub embedded: Option<BitmapFont>,
}

/// Bitmap font data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BitmapFont {
    pub width: u8,
    pub height: u8,
    /// 256 or 512 characters
    pub char_count: u16,
    /// Bitmap data (width * height * char_count bytes)
    pub data: Vec<u8>,
}

/// SAUCE metadata record
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SauceRecord {
    pub title: String,
    pub author: String,
    pub group: String,
    pub date: String,  // CCYYMMDD
    pub columns: u16,  // tinfo1
    pub rows: u16,     // tinfo2
    pub flags: SauceFlags,
    pub font_name: String,
    pub comments: Vec<String>,
}

bitflags! {
    /// SAUCE flags byte
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
    pub struct SauceFlags: u8 {
        const ICE_COLORS         = 0b0000_0001;
        const LETTER_SPACING_9BIT = 0b0000_0100;
        const ASPECT_RATIO       = 0b0000_1000;
    }
}

/// Source format type
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum SourceFormat {
    ANSI,
    Binary,
    PCBoard,
    XBin,
    Tundra,
    ArtWorx,
    IceDraw,
    UTF8ANSI,
}

/// Animation frame
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Frame {
    /// Timestamp in milliseconds
    pub timestamp_ms: u32,
    /// Operations to apply
    pub operations: Vec<Operation>,
}

/// Frame operation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Operation {
    SetCell { x: u32, y: u32, cell: Cell },
    FillRect { x: u32, y: u32, w: u32, h: u32, cell: Cell },
    ScrollRect { x: u32, y: u32, w: u32, h: u32, dx: i32, dy: i32 },
    SetPalette { palette: PaletteType },
}

impl AnsilustIR {
    /// Create a new IR with given dimensions
    pub fn new(width: u32, height: u32) -> Self {
        let size = (width * height) as usize;
        let default_cell = Cell {
            char: b' ' as u32,
            style_id: 0,
            flags: CellFlags::empty(),
        };
        
        let default_style = Style {
            fg: Color::Palette(7),  // White
            bg: Color::Palette(0),  // Black
            underline_color: None,
            attributes: Attributes::empty(),
            hyperlink: None,
        };
        
        Self {
            version: "0.1.0".to_string(),
            width,
            height,
            cells: vec![default_cell; size],
            style_table: vec![default_style],
            grapheme_map: HashMap::new(),
            palette: PaletteType::ANSI,
            font: FontInfo {
                id: Some("cp437".to_string()),
                width: 8,
                height: 16,
                embedded: None,
            },
            sauce: None,
            source_format: SourceFormat::ANSI,
            ice_colors: false,
            letter_spacing: 8,
            aspect_ratio: None,
            frames: None,
        }
    }
    
    /// Get cell at (x, y)
    pub fn get_cell(&self, x: u32, y: u32) -> Option<&Cell> {
        if x >= self.width || y >= self.height {
            return None;
        }
        let index = (y * self.width + x) as usize;
        self.cells.get(index)
    }
    
    /// Set cell at (x, y)
    pub fn set_cell(&mut self, x: u32, y: u32, cell: Cell) -> bool {
        if x >= self.width || y >= self.height {
            return false;
        }
        let index = (y * self.width + x) as usize;
        if let Some(c) = self.cells.get_mut(index) {
            *c = cell;
            true
        } else {
            false
        }
    }
    
    /// Get style by ID
    pub fn get_style(&self, style_id: u16) -> Option<&Style> {
        self.style_table.get(style_id as usize)
    }
    
    /// Add style to table (returns style_id)
    pub fn add_style(&mut self, style: Style) -> u16 {
        // Check if style already exists (for reference counting)
        for (id, existing) in self.style_table.iter().enumerate() {
            if existing == &style {
                return id as u16;
            }
        }
        
        // Add new style
        self.style_table.push(style);
        (self.style_table.len() - 1) as u16
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_create_ir() {
        let ir = AnsilustIR::new(80, 25);
        assert_eq!(ir.width, 80);
        assert_eq!(ir.height, 25);
        assert_eq!(ir.cells.len(), 80 * 25);
    }
    
    #[test]
    fn test_cell_access() {
        let mut ir = AnsilustIR::new(80, 25);
        
        // Set cell
        let cell = Cell {
            char: b'A' as u32,
            style_id: 0,
            flags: CellFlags::empty(),
        };
        assert!(ir.set_cell(0, 0, cell));
        
        // Get cell
        let retrieved = ir.get_cell(0, 0).unwrap();
        assert_eq!(retrieved.char, b'A' as u32);
        
        // Out of bounds
        assert!(ir.get_cell(100, 100).is_none());
        assert!(!ir.set_cell(100, 100, cell));
    }
    
    #[test]
    fn test_style_table() {
        let mut ir = AnsilustIR::new(80, 25);
        
        let style = Style {
            fg: Color::RGB(255, 0, 0),
            bg: Color::None,
            underline_color: None,
            attributes: Attributes::BOLD,
            hyperlink: None,
        };
        
        let id1 = ir.add_style(style.clone());
        let id2 = ir.add_style(style.clone());  // Should reuse
        
        assert_eq!(id1, id2);  // Reference counting
        assert_eq!(ir.style_table.len(), 2);  // Default + new style
    }
    
    #[test]
    fn test_cell_flags() {
        let flags = CellFlags::WIDE_CHAR | CellFlags::SOFT_WRAP;
        assert!(flags.contains(CellFlags::WIDE_CHAR));
        assert!(flags.contains(CellFlags::SOFT_WRAP));
        assert!(!flags.contains(CellFlags::SPACER_TAIL));
    }
}
